File: aps-backend/app/schemas/planning.py
"""Planning schemas for the FE MPS/Work-order views. camelCase serialization
matches the FE contract (orderNo, planQty, workStartDate, ...)."""
from __future__ import annotations

from datetime import date

from pydantic import Field

from app.schemas.master import CamelModel


class MpsOut(CamelModel):
    id: int
    order_no: str | None = Field(None, description="aps_mps_plan.plan_no")
    item_code: str | None = Field(None, description="Item's item_no")
    plan_qty: float | None = Field(None, description="aps_mps_plan.plan_qty")
    end_date: date | None = Field(None, description="delivery_date (customer due date)")
    work_start_date: date | None = Field(None, description="plan_start_date")
    work_end_date: date | None = Field(None, description="plan_end_date")
    status: str = Field("DRAFT", description="Mapped from status_cd: DRAFT | CONFIRMED | CANCELLED")


class WorkOrderOut(CamelModel):
    id: int
    wo_no: str | None = Field(None, description="work_order.work_order_no")
    mps_id: int | None = Field(None, description="work_order.mps_plan_id")
    item_code: str | None = Field(None, description="Item's item_no")
    wc_code: str | None = Field(None, description="Workcenter's workcenter_no")
    plan_qty: float | None = Field(None, description="work_order.qty")
    plan_start_date: date | None = Field(None, description="min(daily_plan.work_date) of the MPS line")
    plan_end_date: date | None = Field(None, description="max(daily_plan.work_date) of the MPS line")
    status: str = Field("PLANNED", description="work_order.status: PLANNED | SENT | CONFIRMED | FAILED")

File: aps-frontend/src/api/planning.ts
import { http } from './http'
import type { Mps, WorkOrder } from '@/types/planning'

// BE `id` int → FE string.
function stringifyId<T extends { id: number | string }>(row: T): T {
  return { ...row, id: String(row.id) } as T
}

// BE `mpsId` cũng int → FE string (nullable).
function stringifyWorkOrder(row: WorkOrder & { mpsId: string | number | null }): WorkOrder {
  return {
    ...row,
    id: String(row.id),
    mpsId: row.mpsId == null ? null : String(row.mpsId),
  }
}

// BE `WorkOrder.status` = PLANNED/SENT/CONFIRMED/FAILED → FE enum PENDING/DONE.
function mapWorkOrderStatus(status: string): 'PENDING' | 'DONE' {
  return status === 'CONFIRMED' ? 'DONE' : 'PENDING'
}

export async function fetchMps(): Promise<Mps[]> {
  const { data } = await http.get<Mps[]>('/planning/mps')
  return data.map(stringifyId)
}

export async function fetchWorkOrders(): Promise<WorkOrder[]> {
  const { data } = await http.get<Array<WorkOrder & { mpsId: string | number | null }>>(
    '/planning/work-orders'
  )
  return data.map((row) => ({
    ...stringifyWorkOrder(row),
    status: mapWorkOrderStatus(row.status),
  }))
}
